### PR TITLE
external modules resolve by node_modules and package.json-main

### PR DESCRIPTION
**Pull request by basarat**
Code for #2338. This is just for the node_modules resolution. Does everything for `node_modules` except for the `typings` stuff as I don't see the rational for that considering there is no way to have tsc send `.js` to one location and `.d.ts` to another location.

[](No longer relevant : 
using `ts.sys.fileExists` doesn't work under the test harness because it doesn't give actual file system paths. So tests are broken by this commit. What is the best way to overcome this. I have reviewed [previous work](https://github.com/Microsoft/TypeScript/compare/master...Nemo157:node_modules) that uses `program.getSourceFile` (which resolves to [this code](https://github.com/Microsoft/TypeScript/blob/e76439b2ff39e1e8afd926450e7cebba2149fe28/src/harness/projectsRunner.ts#L212-L222)) to do exists checks and read `package.json` and that feels *very ugly*. How would you guys recommend authoring the `resolveExternalModule` function.)

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -859,17 +859,9 @@ module ts {
...
-                sourceFile = forEach(supportedExtensions, extension => host.getSourceFile(fileName + extension));
-                if (sourceFile || isRelative) {
-                    break;
-                }
-                let parentPath = getDirectoryPath(searchPath);
-                if (parentPath === searchPath) {
-                    break;
-                }
-                searchPath = parentPath;
+            let resolvedName: string = host.resolveExternalModule(moduleName, searchPath);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Omit the type annotation

Comment (basarat) [inline src/compiler/checker.ts]: Done

**Comment by basarat** [comment]
> How would you guys recommend authoring the resolveExternalModule function.

I've added a commit which gets the tests passing to demonstrate the issue : https://github.com/Microsoft/TypeScript/commit/a7819160fed625aa9d45336ce38d67a79f70ff35

**Comment by DanielRosenwasser** [comment]
Thanks for the PR @basarat - I got the chance to discuss it a bit with @vladima and apparently there are some subtleties on our end, some of which pertain to language service scenarios. Vlad has some existing work on this, and may be able to explain a bit more.

**Comment by basarat** [comment]
> apparently there are some subtleties on our end, some of which pertain to language service scenarios

If it is because the language service didn't get such a resolved file as a part of its original file set : then this can be solved by lazily expanding the file set : https://github.com/TypeStrong/atom-typescript/blob/8df1482222fb88363b504d9c27f98692e938e44f/lib/main/lang/core/languageServiceHost2.ts#L353-L358 I've tried the code in this PR with atom-typescript and it works like a charm :rose:

**Comment by vladima** [comment]
I guess main points of inconvenience on our end are:
1. when language service is hosted inside the VS you in general don't have access to all `sys` functionality meaning that operations like `isDirectoryExists`, `isFileExists` and `readFile` should be delegated to the host.
2. the way how VS side works today is slightly different from the model used by tsserver. Main difference is that currently VS pre-resolves all tripleslash-refs/imported files before creating script language service so if resolution process will be just changed on the script side - this will break VS integration. 
3. per #2338 we would like to have ability to switch module resolution from node style (with folder walks) to requirejs (that will use baseURL) and vice versa. Ideally module resolver should just a function `(moduleName: string, containingFile: string, compilerHost: CompilerHost) => string` so if somebody needs more involved method to resolve modules - he can do this by just replacing this one part. It does not directly contradict with proposed solution - it just needs to be slightly reshaped.

**Comment by basarat** [comment]
> I guess main points of inconvenience

@vladima Taking your points into consideration I guess you are recommending:

* adding `readFile` and `fileExists` to host ... and then using `sys` inside host. 
* after that refactoring use the resolver function to be `(moduleName: string, containingFile: string, compilerHost: CompilerHost) => string` and use `host.readFile`/`host.fileExists` inside. 

So:

* ^ Is the refactoring I am suggestion correct?
* What is shim host? < you might same me some time digging through the code :)

**Comment by vladima** [comment]
> What is shim host?

do you mean [this guy](https://github.com/Microsoft/TypeScript/blob/master/src/services/shims.ts#L46) ? The idea behind it (and shims in general) is that they act as a send\receive proxies whose purpose is to extract incoming parameters from JSON, call corresponding method on the language service instance, wrap its return value to JSON and send it back.

>  Is the refactoring I am suggestion correct?

In general yes, you got the idea. I need to think more about details of interaction with VS that should be handled.

**Comment by basarat** [comment]
> do you mean

Yup. Thanks. I'll see if it comes up in this refactoring (I am sure it will) :)

> In general yes, you got the idea

I'll give it a go :rose:

**Comment by basarat** [comment]
> Investigating a better method

Pushed https://github.com/basarat/TypeScript/commit/a17c6aaec3b25624111c3fcd8026b5af2340c471 

The  [current changeset is ready for an early review if anyone is interested](https://github.com/Microsoft/TypeScript/pull/3147/files). I'm moving on to adding tests for this.

**Comment by basarat** [comment]
Tests pushed. The PR is ready for review :+1: :rose:

**Comment by Zoltu** [comment]
I don't believe this technique is a good idea for large application development.  Auto-detection is nice for toy projects (hello world apps, scripts, etc.) but has potential for causing headaches and very difficult to track down bugs in large applications.

An example would be if you bring in a new node module that happens to share a name with an existing library you (or someone else) wrote and is resulting in a name collision.  In this case you will end up pulling in the wrong typings when importing without realizing it.  If both imported types have similar structure it further complicates the matter because things will look *mostly* right.

Also, this would need to filter out devDependencies.  If I have `underscore` as a dev dependency and import it with `import underscore from 'underscore';` TSC will find the dev dependency and bring in the typings, telling me everything compiled fine.  However, when I go to deploy underscore won't be bundled/deployed because it was listed as a devDependency.  If lucky, it wouldn't build on the build server... if not so lucky you'll just have production builds that fail to work properly.

Personally, I am a much bigger fan of an explicit mapping in `tsconfig.json`.  While people may be turned off by the idea of having to populate that file by hand, it should really be auto-generated by tooling in most cases (see JSPM, Visual Studio C#, etc.).  The explicit mappings makes troubleshooting easier because there is no "magic" mapping behind the scenes at compile time, the magic happens at tooling time.  I believe it is also easier to document behavior of.

If it is decided to keep auto-mapping like this, I would recommend a configuration flag that can disable it.

**Comment by Zoltu** [comment]
Ah, I should have put these comments over in #2338.  For some reason I thought this PR was for #2839, I must have got them mixed up in my head.  Will leave comment above to avoid confusion for people who already received email notification, feel free to delete it if desirable.

**Comment by Zoltu** [comment]
TSC may not be able to output a d.ts to a separate location, but a post-compile build step easily could.  It seems to be fairly common for typings to sit in a typings directory.

That being said, I believe there is value in setting a standard for how people *should* distribute their TS/JS packages (no typings folder) and expand it later if necessary.

**Comment by basarat** [comment]
> believe there is value in setting a standard for how people should distribute their TS/JS packages (no typings folder) and expand it later if necessary.

Thanks :heart: 

Also at the moment the typings from the `typings` property are currently of the form `declare module "foo/bar"` and those typings are not the target of this PR anyways :rose:

**Comment by poelstra** [comment]
:+1: on this functionality!

**Comment by vladima** [comment]
I've pulled with PR into the [moduleResolution branch](https://github.com/Microsoft/TypeScript/tree/moduleResolution) in this repo - this will cover node related part of #2338 and the rest of the work on this proposal will be continued in that branch.

Thanks!

**Comment by basarat** [comment]
Thanks @vladima!

**Comment by mhegazy** [comment]
This should be in the 1.6 release

**Pull request closed without merging**